hpx/lcos/detail/promise_base.hpp

```diff
@@ -0,0 +1,222 @@
+              , id_retrieved_(false)
+            {
+                // The lifetime of the LCO (component) part is completely
+                // handled by the shared state, we create the object to get our
+                // gid and then attach it to the completion handler of the
+                // shared
+                // state.
+                typedef std::unique_ptr<wrapping_type> wrapping_ptr;
+                wrapping_ptr                           lco_ptr(
+                    new wrapping_type(new wrapped_type(this->shared_state_)));
+
+                id_   = lco_ptr->get_unmanaged_id();
+                addr_ = naming::address(hpx::get_locality(),
+                    lco_ptr->get_component_type(),
+                    lco_ptr.get());
+
+                // This helper is used to keep the component alive until the
+                // completion handler has been called.
+                auto keep_alive = hpx::util::deferred_call(
+                    [](wrapping_ptr ptr) {}, std::move(lco_ptr));
+                this->shared_state_->set_on_completed(std::move(keep_alive));
+            }
+
+            promise_base(promise_base&& other) HPX_NOEXCEPT
+                : base_type(std::move(other)),
+                  id_retrieved_(other.id_retrieved_),
+                  id_(std::move(other.id_)),
+                  addr_(std::move(other.addr_))
+            {
+                other.id_   = naming::invalid_id;
+                other.addr_ = naming::address();
+            }
+
+            ~promise_base()
+            {
+                check_abandon_shared_state(
+                    "lcos::detail::promise_base<R>::~promise_base()");
+                this->shared_state_.reset();
+            }
+
+            promise_base& operator=(promise_base&& other) HPX_NOEXCEPT
+            {
+                base_type::operator=(std::move(other));
+                id_retrieved_      = other.id_retrieved_;
+                id_                = std::move(other.id_);
+                addr_              = std::move(other.addr_);
+
+                other.id_   = naming::invalid_id;
+                other.addr_ = naming::address();
+                return *this;
+            }
+
+            naming::id_type get_id(error_code& ec = throws) const
+            {
+                if (this->shared_state_ == 0)
+                {
+                    HPX_THROWS_IF(ec, no_state,
+                        "detail::promise_base<Result, RemoteResult>::get_id",
+                        "this promise has no valid shared state");
+                    return naming::invalid_id;
+                }
+                if (!addr_ || !id_)
+                {
+                    HPX_THROWS_IF(ec, no_state,
+                        "detail::promise_base<Result, RemoteResult>::get_id",
+                        "this promise has no valid LCO");
+                    return naming::invalid_id;
+                }
+                if (!this->future_retrieved_)
```

Comment:
Do you still have to ensure that the future is retrieved before the `get_id()` was called? Or is that a left-over?